community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreVersionCheck.java

```diff
@@ -82,24 +82,32 @@ private String typeDescriptor( String expectedVersion )
         return expectedVersion.substring( 0, spaceIndex );
     }
 
-    private String readVersion( PagedFile file, int charsToRead ) throws IOException
+    private String readVersion( PagedFile pagedFile, String expectedVersion ) throws IOException
     {
+        int charsToRead = expectedVersion.length();
+        int maximumNumberOfPagesVersionSpans = UTF8.encode( expectedVersion ).length / pagedFile.pageSize() + 2;
         String version = null;
-        try ( PageCursor cursor = file.io( file.getLastPageId(), PagedFile.PF_SHARED_LOCK ) )
+        try ( PageCursor pageCursor = pagedFile.io(
+                Math.max( pagedFile.getLastPageId() + 1 - maximumNumberOfPagesVersionSpans, 0 ),
+                PagedFile.PF_SHARED_LOCK ) )
         {
-            if ( cursor.next() )
+            int currentPage = 0;
+            byte[] allData = new byte[pagedFile.pageSize() * maximumNumberOfPagesVersionSpans];
+            while ( pageCursor.next() )
             {
-                byte[] data = new byte[cursor.getCurrentPageSize()];
+                byte[] data = new byte[pagedFile.pageSize()];
                 do
                 {
-                    cursor.getBytes( data );
-                }
-                while ( cursor.shouldRetry() );
-                String pageContents = UTF8.decode( data ).trim();
-                if ( pageContents.length() >= charsToRead )
-                {
-                    version = pageContents.substring( pageContents.length() - charsToRead );
+                    pageCursor.getBytes( data );
                 }
+                while ( pageCursor.shouldRetry() );
+                System.arraycopy( data, 0, allData, currentPage * data.length, data.length );
+                currentPage++;
+            }
+            String pageContents = UTF8.decode( allData ).trim();
```

Comment:
I don't think you can just UTF-8 decode this big blob of binary, because in UTF-8 some bytes dictate how subsequent bytes are interpreted. You need to look for the binary pattern that is the version. There's also the issue that you cannot be guaranteed to get a consistent read across two pages, though in practice I don't think that will be a problem, because this utility should never be used while the file is otherwise mapped for other purposes.